### PR TITLE
gh-64662: Add virtual table support to sqlite3.Connection.iterdump

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-64662

----
:books: Documentation preview :books:: https://cpython-previews--108340.org.readthedocs.build/

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@felixxm, would you like to review this? This is pretty much @palaviv's patch, adapted to `main` by me.

**Comment by erlend-aasland** [comment]
Further experiments show that `PRAGMA foreign_keys=OFF;` is prepended, no matter if virtual tables are involved or not, so we should fix that separately IMO.

**Comment by felixxm** [comment]
Maybe it's just me, but the number of different quotes in `_iterdump()` makes it hard for me to understand this code, check out #108472 (feel-free to reject it).

File: Lib/sqlite3/dump.py [hunk 1]
```diff
@@ -50,13 +51,15 @@ def _iterdump(connection):
...
-        #    qtable = table_name.replace("'", "''")
-        #    yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"\
-        #        "VALUES('table','{0}','{0}',0,'{1}');".format(
-        #        qtable,
-        #        sql.replace("''")))
+        elif sql.startswith('CREATE VIRTUAL TABLE'):
+            if not writeable_schema:
+                writeable_schema = True
+                yield('PRAGMA writable_schema=ON;')
+            qtable = table_name.replace("'", "''")
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
            qtable = _quote_value(table_name)
```

File: Lib/sqlite3/dump.py [hunk 2]
```diff
@@ -50,13 +51,15 @@ def _iterdump(connection):
...
+            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
+                  "VALUES('table','{0}','{0}',0,'{1}');".format(
+                    qtable,
+                    sql.replace("'", "''")))
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
                    _quote_value(sql)))
```

File: Lib/sqlite3/dump.py [hunk 3]
```diff
@@ -50,13 +51,15 @@ def _iterdump(connection):
...
-        #        sql.replace("''")))
+        elif sql.startswith('CREATE VIRTUAL TABLE'):
+            if not writeable_schema:
+                writeable_schema = True
+                yield('PRAGMA writable_schema=ON;')
+            qtable = _quote_value(table_name)
+            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
+                  "VALUES('table','{0}','{0}',0,'{1}');".format(
+                    qtable,
+                    _quote_value(sql)))
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
                  "VALUES('table','{0}','{0}',0,'{1}');".format(
                      qtable,
                      _quote_value(sql),
                  ))
```

File: Lib/sqlite3/dump.py [hunk 4]
```diff
@@ -50,13 +51,16 @@ def _iterdump(connection):
...
+        elif sql.startswith('CREATE VIRTUAL TABLE'):
+            if not writeable_schema:
+                writeable_schema = True
+                yield('PRAGMA writable_schema=ON;')
+            qtable = _quote_value(table_name)
+            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
+                  "VALUES('table','{0}','{0}',0,'{1}');".format(
+                      qtable,
+                      _quote_value(sql),
+                  ))
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
                  "VALUES('table','{0}','{0}',0,'{1}');".format(
                      _quote_value(table_name),
                      _quote_value(sql),
                  ))
```

File: Lib/sqlite3/dump.py [hunk 5]
```diff
@@ -50,13 +51,15 @@ def _iterdump(connection):
...
-        #    yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"\
-        #        "VALUES('table','{0}','{0}',0,'{1}');".format(
-        #        qtable,
-        #        sql.replace("''")))
+        elif sql.startswith('CREATE VIRTUAL TABLE'):
+            if not writeable_schema:
+                writeable_schema = True
+                yield('PRAGMA writable_schema=ON;')
+            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
+                  "VALUES('table','{0}','{0}',0,'{1}');".format(
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
                  "VALUES('table',{0},{0},'0',{1});".format(
```

File: Lib/sqlite3/dump.py [hunk 6]
```diff
@@ -50,13 +51,15 @@ def _iterdump(connection):
...
-        #    yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"\
-        #        "VALUES('table','{0}','{0}',0,'{1}');".format(
-        #        qtable,
-        #        sql.replace("''")))
+        elif sql.startswith('CREATE VIRTUAL TABLE'):
+            if not writeable_schema:
+                writeable_schema = True
+                yield('PRAGMA writable_schema=ON;')
+            yield("INSERT INTO sqlite_master(type,name,tbl_name,rootpage,sql)"
+                  "VALUES('table',{0},{0},'0',{1});".format(
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]:
```suggestion
                  "VALUES('table',{0},{0},0,{1});".format(
```

**Review by felixxm — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks for the review, Mariusz!

**Pull request merged**